### PR TITLE
Fix our "foreign" handling (GOOS and GOARCH that aren't GOHOST*)

### PR DESCRIPTION
**Pull request by tianon**

### CODE REVIEW

**Comment by tianon** [comment]
This one is kind of hard to test, but I feel like we should.  I'll see what I can come up with.

**Comment by tianon** [comment]
Oh, we kind of are with those `arm` tests, but we were getting false positives because it wasn't setting `GOARCH` correctly.

**Comment by tianon** [comment]
Ok, I've updated the `.travis.yml` to do even more foreign testing (`linux/arm` on both Linux and OSX), including verifying at least at a basic level that it worked appropriately.

File: gimme [hunk 1]
```diff
@@ -158,18 +159,24 @@ _extract() {
...
-			"${GIMME_VERSION_PREFIX}/go1.4"*".${GIMME_OS}.${GIMME_HOSTARCH}"
-		)
-		if [ "${#bootstrap_candidates[@]}" -gt 0 ]; then
-			bootstrap_root="$bootstrap_candidates"
-		fi
-		if [ -z "${bootstrap_root}" ] ; then
-			echo >&2 "error: Missing bootstrap installation"
-			echo >&2 "error: Unable to compile go from git clone"
-			return 1
+		if grep -q GOROOT_BOOTSTRAP "${1}/src/make.bash" &> /dev/null ; then
```
Comment (meatballhat) [inline gimme:162]: nice!

File: gimme [hunk 2]
```diff
@@ -177,9 +184,8 @@ _compile() {
 		fi
 		cd src
 		export GOOS="${GIMME_OS}" GOARCH="${GIMME_ARCH}"
-		export GOROOT_BOOTSTRAP="${bootstrap_root}"
 		local make_log="${1}/make.${GOOS}.${GOARCH}.log"
-		./make.bash &> $make_log
+		./make.bash &> $make_log || return 1
```
Comment (meatballhat) [inline gimme:188]: :facepalm:

File: gimme [hunk 3]
```diff
@@ -196,11 +202,11 @@ EOF
 
 # _env "dir"
 _env() {
-	[ -d "${1}/bin" -a -x "${1}/bin/go" ] || return
+	[ -d "${1}/bin" -a -x "${1}/bin/go" ] || return 1
```
Comment (meatballhat) [inline gimme:205]: YES

File: gimme [hunk 4]
```diff
@@ -196,11 +202,11 @@ EOF
...
 
 	# if we try to run a Darwin binary on Linux, we need to fail so 'auto' can fallback to cross-compiling from source
 	# automatically
-	GOROOT="${1}" "${1}/bin/go" version &> /dev/null || return
+	GOROOT="${1}" "${1}/bin/go" version &> /dev/null || return 1
```
Comment (meatballhat) [inline gimme:209]: MUCH YES

**Comment by tianon** [comment]
OMG

**Comment by tianon** [comment]
:green_heart:

**Comment by tianon** [comment]
BTW, this changes part of how `~/.gimme/envs` works.  Previously, if I did `GIMME_OS=darwin GIMME_ARCH=386 ./gimme 1.4.1` on `linux/amd64`, it would create `go1.4.1.src.env` because we have to build from source to get that cross-compile combination to work.  Now it instead creates `go1.4.1.darwin.386.env` which points to `~/.gimme/versions/go1.4.1.src` as `GOROOT` since that directory now is the canonical place to find a successfully compiled copy of `darwin/386`.

**Comment by tianon** [comment]
Ok @meatballhat, definitely ready for proper review now. :+1:

**Pull request merged**